crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2156,108 +2115,100 @@ impl<'db> Type<'db> {
                                 binding.set_return_type(len_ty);
                             }
                         };
-
-                        CallOutcome::callable(binding)
                     }
 
                     Some(KnownFunction::Repr) => {
                         if let Some(first_arg) = binding.one_parameter_type() {
                             binding.set_return_type(first_arg.repr(db));
                         };
-
-                        CallOutcome::callable(binding)
-                    }
-
-                    Some(KnownFunction::AssertType) => {
-                        let Some((_, asserted_ty)) = binding.two_parameter_types() else {
-                            return CallOutcome::callable(binding);
-                        };
-
-                        CallOutcome::asserted(binding, asserted_ty)
                     }
 
                     Some(KnownFunction::Cast) => {
                         // TODO: Use `.two_parameter_tys()` exclusively
                         // when overloads are supported.
-                        if binding.two_parameter_types().is_none() {
-                            return CallOutcome::callable(binding);
-                        };
-
                         if let Some(casted_ty) = arguments.first_argument() {
-                            binding.set_return_type(casted_ty);
+                            if binding.two_parameter_types().is_some() {
+                                binding.set_return_type(casted_ty);
+                            }
                         };
-
-                        CallOutcome::callable(binding)
                     }
 
-                    _ => CallOutcome::callable(binding),
+                    _ => {}
```

Comment:
I think I'd prefer an exhaustive match here, since it seems like we special-case the return-type for a majority of the `KnownFunction` variants (at least right now):

```suggestion
                    // These remaining functions do not have the inferred return-type special-cased in any way.
                    Some(
                        KnownFunction::ConstraintFunction(_)
                        | KnownFunction::Final
                        | KnownFunction::RevealType
                        | KnownFunction::NoTypeCheck
                        | KnownFunction::StaticAssert
                        | KnownFunction::AssertType,
                    )
                    | None => {}
```